File: water_edison/src/main/java/com/viomi/waterpurifier/edison/util/SchedulerProvider.java
package com.viomi.waterpurifier.edison.util;

import androidx.annotation.NonNull;

import io.reactivex.ObservableTransformer;
import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * @description:
 * @data:2021/12/15
 */
public class SchedulerProvider {
    private static SchedulerProvider instance;

    public static synchronized SchedulerProvider getInstance() {
        if (instance == null) {
            Class var0 = SchedulerProvider.class;
            synchronized (SchedulerProvider.class) {
                instance = new SchedulerProvider();
            }
        }

        return instance;
    }

    private SchedulerProvider() {
    }

    @NonNull
    public Scheduler computation() {
        return Schedulers.computation();
    }

    @NonNull
    public Scheduler io() {
        return Schedulers.io();
    }

    @NonNull
    public Scheduler ui() {
        return AndroidSchedulers.mainThread();
    }

    @NonNull
    public <T> ObservableTransformer<T, T> applySchedulers() {
        return (observable) -> {
            return observable.subscribeOn(this.io()).observeOn(this.ui());
        };
    }
}

File: common_upgrade/src/main/java/com/viomi/upgrade_lib/utils/SchedulerProvider.kt
package com.viomi.upgrade_lib.utils

import io.reactivex.Observable
import io.reactivex.ObservableTransformer
import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

/**
 * Created by mai on 2020/12/2.
 */
class SchedulerProvider {


    companion object {
        fun io(): Scheduler? {
            return Schedulers.io()
        }

        fun ui(): Scheduler? {
            return AndroidSchedulers.mainThread()
        }

        fun <T> applySchedulers(): ObservableTransformer<T, T>? {
            return ObservableTransformer { observable: Observable<T> ->
                observable.observeOn(
                    ui()
                ).subscribeOn(io())
            }
        }

        fun <T> applySchedulers2(): ObservableTransformer<T, T>? {
            return ObservableTransformer { observable: Observable<T> ->
                observable.observeOn(
                    io()
                ).subscribeOn(io())
            }
        }
    }

}